### DISCUSSION TITLE
Move to LUA

### DISCUSSION THREAD

**Discussion by jrock2004** [nicknisi/dotfiles#173] [Ideas]
I have started the transition but wondering if others have moved over to LUA. Did you like it? What issues did you have? Or maybe you think its a bad idea?

**Comment by yarthur** [comment]
I’m not one of the hip kids, so if you would, please explain a bit further; are you thinking of using LUA in place of the shell scripts, as a lifestyle choice, or something else altogether?

My shell scripting knowledge is almost entirely StackOverflow based. It’s what bothers me the most about my setup scripts; I barely know how they work _at best_. If LUA offers something a bit less inscrutable to my limited experience, that would be very interesting to me.

**Reply by nicknisi** [reply]
Neovim has Lua [built-in](https://neovim.io/doc/user/lua.html) and it can be used as a way to script the editor as an alternative to Vim script. I haven't really considered it yet because it's always seemed like it might move me too far from actual Vim in case I ever wanted to make the switch back. That being said, I see a lot of folks making this switch. 

As for shell scripting, I kind of enjoy writing Bash. The problem I have with writing in other languages is that you need to ensure the runtime is there, along with any other dependencies. I believe one benefit of either Go or Deno is that you can compile an executable that contains the runtime, though, so this could be a good option.

**Reply by jrock2004** [reply]
Another reason to move over is that it support language servers without the need for plugins to do it. Here is an example of having inline type error messaging.

![image](https://user-images.githubusercontent.com/655716/112768580-a3e8e300-8fea-11eb-8027-3f3541e70025.png)

I think lua looks more like a programming language to me them Vimscript. Still learning things but its been fun in my move over to LUA

**Reply by nicknisi** [reply]
@jrock2004 oh wow, is that just for the Lua scripts?

**Reply by jrock2004** [reply]
Nope

**Reply by jrock2004** [reply]
Here is some errors in typescript

![image](https://user-images.githubusercontent.com/655716/112882384-e538cb80-909a-11eb-9b44-369fa669175c.png)

**Reply by jrock2004** [reply]
And here is some more good LSP goodness

![image](https://user-images.githubusercontent.com/655716/112883113-d30b5d00-909b-11eb-920d-b7b0e5f22ac2.png)

**Reply by nicknisi** [reply]
Ok, now I'm really interested in this. I have mostly ignored it as I've gone down the [coc](https://github.com/neoclide/coc.nvim) route. But, it's the switch to a Lua config that enables this?

**Reply by jrock2004** [reply]
I am still learning but yeah. There are even plugins that will change for you. Treesitter is another nice thing as you get better syntax highlighting. For best results I switched to neovim 0.5.5. Its 97% done and almost ready to switch. Its definitely an investment. Not a clone of your repo, but here is where I am working on the conversion https://github.com/jrock2004/dotfiles/tree/convert-to-lua/files/.config/nvim

@nicknisi was even able to replace fzf with a lua plugin

![image](https://user-images.githubusercontent.com/655716/112885251-87a67e00-909e-11eb-9ff1-21f17f613734.png)

**Reply by yarthur** [reply]
In case I came off too snarky or otherwise off-putting, my apologies. And of course, I dropped a comment, and then ghosted for a couple weeks 🤦‍♂️

> Neovim has Lua [built-in](https://neovim.io/doc/user/lua.html) and it can be used as a way to script the editor as an alternative to Vim script.

I was genuinely clueless about this, so thanks for explaining it.

**Comment by jrock2004** [comment]
So have I convinced anyone to switch to Lua config?

**Reply by nicknisi** [reply]
This is something I'm interested in learning more about! If you need to learn a language to configure your editor, Lua is probably better than vim script, but there's also a lot to convert.

**Reply by jrock2004** [reply]
Well that is nice is that you do not get too deep into Lua. You won't need to learn much beyond the basics. It was definitely an investment to get my dot files converted

**Reply by nicknisi** [reply]
Do you have any resources that you used to learn? It seems like a lot of it would be "How do I do <vim script thing> in Lua?" searches.

**Reply by jrock2004** [reply]
There is a YouTube series where someone was converting their complex Vimscript setup to Lua. I watched to get an understanding then I just started to convert mine. https://www.youtube.com/watch?v=BoDU6QOmEOY

**Reply by jrock2004** [reply]
But you are right that I should go find some tutorials on lua to get some more understanding

**Reply by jrock2004** [reply]
https://www.lua.org/pil/contents.html Its pretty good

**Comment by jrock2004** [comment]
One sad thing about moving to LUA is that I get farther away from using this repo as it has helped me so much in the past

**Reply by nicknisi** [reply]
I definitely intend to check out your config!

**Comment by nicknisi** [comment]
I have finally started taking a look at this. Initially, I have some very basic Lua set up to configure the LSP and I removed coc. The work is on [this branch](https://github.com/nicknisi/dotfiles/tree/lua-neovim-lsp). So far, I am really excited about all you can do with these new plugins!

**Reply by jrock2004** [reply]
Yeah you really need to look back and see what plugins can replace what. There is some M1 problems with some things. I am still working and some things are broken for me

**Reply by nicknisi** [reply]
Ooh, I am very curious what M1 problems you're seeing? So far everything has been flawless on mine.

**Reply by jrock2004** [reply]
lua language server

**Reply by jrock2004** [reply]
Oooh, looks like it might be fixed

**Reply by jrock2004** [reply]
https://github.com/Koihik/LuaFormatter/issues/180 This is the issue to watch

**Comment by nicknisi** [comment]
Continuing this thread, I have a branch with my `init.vim` ported to `init.lua` https://github.com/nicknisi/dotfiles/blob/lua-neovim-lsp/config/nvim/init.lua

**Reply by nicknisi** [reply]
It's pretty straightforward for the most part, but there are a few things that aren't working that I'm still looking in to.

**Reply by jrock2004** [reply]
I think the 2 things that got me was the settings and knowing how to do them. The auto commands are a little bit of a pain

**Reply by nicknisi** [reply]
Yeah, I agree. I mostly just fall back to `vim.cmd` and `vim.api.nvim_exec` calls for stuff like that. I think I am at parity with my init.vim now, which is awesome!

**Comment by jrock2004** [comment]
I will say that my setup is not fully right and I need to find some time to fix some things. The things that are working are nice

**Reply by nicknisi** [reply]
I'd love to help understand what's not working well. I think I am back up to parity with where my init.vim was and have started adding some cool new Lua plugins like [trouble](https://github.com/folke/trouble.nvim)!

**Reply by jrock2004** [reply]
Yeah, when I decided to start this journey, I decided to replace my plugins with lua plugins and use LSP. I still use old vim plugins, but only when needed

**Reply by jrock2004** [reply]
I just installed trouble and man, this is a really cool plugin

**Comment by jrock2004** [comment]
![image](https://user-images.githubusercontent.com/655716/127726927-847a7db8-8297-4911-9500-fd9fd1d0dc14.png)

I am using a cool dashboard plugin that opens on first load of vim. When I select something or open a file with telescope, I would love for it to close that buffer. This is the first problem I need to fix this weekend

**Reply by nicknisi** [reply]
Is it [this plugin](https://github.com/glepnir/dashboard-nvim) you're using? I looked at that briefly and am wondering about replacing startify with it.

**Reply by jrock2004** [reply]
Yeah, I am using someones fork of it, but yes https://github.com/ChristianChiarulli/dashboard-nvim

**Reply by jrock2004** [reply]
I fixed the issue from switching from using the fork to using the original

**Comment by jrock2004** [comment]
Problem 2 I have is when I save some files that does formatting on save, and it puts it as changes so I see the circle. I save file again and it does same thing. If I hit `:q` I get a warning that there is unsaved changes

![image](https://user-images.githubusercontent.com/655716/127747723-ca55c656-1b51-4434-9ec5-be407a86e0d3.png)

Here is the section that does the formatting

```
vim.cmd [[augroup Format]]
vim.cmd [[autocmd! * <buffer>]]
vim.cmd [[autocmd BufWritePost <buffer> lua formatting()]]
vim.cmd [[augroup END]]
```

**Reply by jrock2004** [reply]
This is fixed now :)

**Comment by jrock2004** [comment]
Just found out with telescope you can use fuzzy search to look up help stuff

![image](https://user-images.githubusercontent.com/655716/127757967-66e43e21-58c0-4928-b621-0ea3ffa07280.png)

```lua
nmap('<leader>vh', '<cmd>lua require(\'telescope.builtin\').help_tags()<CR>')
```
